File: HwangSettong/week_3/177p.java
// https://school.programmers.co.kr/learn/courses/30/lessons/81303
import java.util.*;
class Solution {
    public String solution(int n, int k, String[] cmd) {
        StringBuilder answer = new StringBuilder();
        Stack<Integer> stack = new Stack<>();
        int[][] list = new int[n][2]; // {prev, next}
        int temp;
        for(int i = 0; i<n; i++){
            list[i][0] = i-1; // prev
            list[i][1] = i+1; // next
            answer.append("O");
        }
        for(String c : cmd){
            if(c.charAt(0) == 'D'){
                temp = Integer.parseInt(c.split(" ")[1]);
                for(int i = 0; i<temp; i++){
                    k = list[k][1];
                }
            }else if(c.charAt(0) == 'U'){
                temp = Integer.parseInt(c.split(" ")[1]);
                for(int i = 0; i<temp; i++){
                    k = list[k][0];
                }
            }else if(c.charAt(0) == 'C'){
                stack.push(k);
                //list[prev]의 next값을 현재 next값으로 변경
                if(list[k][0] > -1){
                    list[list[k][0]][1] = list[k][1];
                }
                //list[next]의 prev값을 현재 prev값으로 변경
                if(list[k][1] < n){
                    list[list[k][1]][0] = list[k][0]; 
                }
                // k 값 변경
                k = list[k][1] < n ? list[k][1] : list[k][0];
            }else{
                temp = stack.pop();
                if(list[temp][1] < n){
                    list[list[temp][1]][0] = temp;
                }if(list[temp][0] > -1){
                    list[list[temp][0]][1] = temp;
                }
            }
        }
        while(!stack.isEmpty()){
            temp = stack.pop();
            answer.replace(temp, temp+1, "X");
        }
        
        return answer.toString();
    }
}

File: jane-kim-279/week_3/177p.py
# https://school.programmers.co.kr/learn/courses/30/lessons/81303
# 효율성 통과X
# linked list로 하면 될까..?

def solution(n, k, cmd):
    table = list(range(n))
    deleted = ['O'] * n
    stack = []
    current = k

    for c in cmd:
        if c[0] == 'U':
            current -= int(c[2:])
        elif c[0] == 'D':
            current += int(c[2:])
        elif c[0] == 'C':
            deleted[table[current]] = 'X'
            stack.append(table[current])
            table.pop(current)
            if current == len(table):
                current -= 1
        elif c[0] == 'Z':
            restored = stack.pop()
            deleted[restored] = 'O'
            if restored < table[current]:
                current += 1
            insert_index = 0
            for i, value in enumerate(table):
                if value > restored:
                    insert_index = i
                    break
            else:
                insert_index = len(table)
            table.insert(insert_index, restored)

    return ''.join(deleted)
